arrow_struct/src/lib.rs

```diff
@@ -11,17 +11,60 @@ use arrow::datatypes::{
 pub use arrow::record_batch::RecordBatch;
 pub use bytes::Bytes;
 use std::fmt::Debug;
+pub use std::option::Option;
 
 pub use arrow_struct_derive::Deserialize;
 
 pub trait FromArrayRef<'a>: Sized {
     fn from_array_ref(array: &'a ArrayRef) -> impl Iterator<Item = Self>;
 }
 
+pub trait FromArrayRefOpt<'a>: Sized {
+    type Item;
+    fn from_array_ref_opt(array: &'a ArrayRef) -> impl Iterator<Item = Option<Self::Item>>;
+}
+
+impl<'a, T: FromArrayRefOpt<'a, Item = T>> FromArrayRef<'a> for Option<T> {
+    fn from_array_ref(array: &'a ArrayRef) -> impl Iterator<Item = Self> {
+        T::from_array_ref_opt(array)
+    }
+}
+
+impl<'a, T: FromArrayRefOpt<'a, Item = T>> FromArrayRefOpt<'a> for Option<T> {
+    type Item = T;
+
+    fn from_array_ref_opt(array: &'a ArrayRef) -> impl Iterator<Item = Option<Self::Item>> {
+        T::from_array_ref_opt(array)
+    }
+}
+
+// Effectively a marker trait, since stable Rust does not have specialization or negative trait bounds
```

Comment:
I'm not sure exactly what you would do with specialization, but have you considered [autoref specialization](https://github.com/dtolnay/case-studies/blob/master/autoref-specialization/README.md)?